src/frontend/src/optimizer/plan_node/logical_kafka_scan.rs

```diff
@@ -0,0 +1,345 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::cmp::{max, min};
+use std::ops::Bound;
+use std::ops::Bound::{Excluded, Included, Unbounded};
+use std::rc::Rc;
+
+use pretty_xmlish::{Pretty, XmlNode};
+use risingwave_common::catalog::{Schema, KAFKA_TIMESTAMP_COLUMN_NAME};
+use risingwave_connector::source::DataType;
+
+use super::generic::GenericPlanRef;
+use super::utils::{childless_record, Distill};
+use super::{
+    generic, BatchProject, ColPrunable, ExprRewritable, Logical, LogicalFilter, LogicalProject,
+    PlanBase, PlanRef, PredicatePushdown, ToBatch, ToStream,
+};
+use crate::catalog::source_catalog::SourceCatalog;
+use crate::error::Result;
+use crate::expr::{Expr, ExprImpl, ExprRewriter, ExprType, ExprVisitor};
+use crate::optimizer::plan_node::expr_visitable::ExprVisitable;
+use crate::optimizer::plan_node::utils::column_names_pretty;
+use crate::optimizer::plan_node::{
+    BatchKafkaScan, ColumnPruningContext, LogicalSource, PredicatePushdownContext,
+    RewriteStreamContext, ToStreamContext,
+};
+use crate::utils::{ColIndexMapping, Condition};
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct LogicalKafkaScan {
+    pub base: PlanBase<Logical>,
+    pub core: generic::Source,
+
+    /// Expressions to output. This field presents and will be turned to a `Project` when
+    /// converting to a physical plan, only if there are generated columns.
+    output_exprs: Option<Vec<ExprImpl>>,
```

Comment:
Random thought: having `output_exprs` here is a little sign of copy-pasting to me. Maybe we can let `LogicalSource::create` directly return `Project + Source`. (Maybe not in this PR)

Reply:
Agree. Let's do it later. `output_exprs` could easily introduce a potential bug, for example, the schema of the `LogicalSource` could be different from its `core` schema which is caught by the equivalent plan checking of the optimizer during the development of this PR.